### ISSUE TITLE
Enable auto builds on local deploy

### ISSUE THREAD

**Issue by lkingland** [knative-sandbox/kn-plugin-func#1075]
## Summary

The `--build=auto` option recently added to the `run` subcommand should also be added to `deploy`.

In order to support this complexity, and two additional scenarios (discussed below) I would suggest we separate the current deploy subcommand's flags into two:  `--remote` indicating the deployment should not be done locally but should instead trigger a remote on-cluster deployment and `--build=[true|false|auto]` to match that which is currently on `run`.

## Explication

Currently the `--build` flag on `deploy` is used to both indicate if building should take place as well as whether the deployment should be triggered remotely on cluster rather than locally (`--build=[local|disabled|git]`).  We currently support building and deploying Functions 1) from the local system or 2) by triggering a build and deploy via Tekton pipelines.

With the introduction of cached builds (See #1031) and its `--build=auto|true|false`), the way these dimensions are expressed as flags becomes more complex.  Since we may soon add the third dimension of git repository vs source tarball (currently we only support git repository remotely and tarball locally), it is even more important to be clear before permutations start to expand.

For example, triggering a remote deploy but with a forced rebuild, or a remote deploy using a tarball of local source code, or a local deploy using source from a remote git repository are all scenarios that should in theory be possible but would be onerous when expressed as a single flag.  Allowing all permutations via a single `--build` flag worked until now, because we only had three possible scenarios, but now with the addition of `--build=auto|true|false`, it seems clear separation is in order.

The proposal is to simply separate into two flags:

`--remote`: whether or not we are deploying using local processing or triggering a remote deployment on-cluster.  Default is local, but will be configurable both on a per-Function basis, and globally as a user setting when the config system is introduced.

`--build`: whether a build should be performed, skipped or performed if the system detects it is necessary (`--build=auto`).

See #1031 for an example of how the `--build` flag is implemented on the `run` subcommand, which is intended as a trial run for how it would work on `deploy`.

### Examples:

```
func deploy
```
Perform a deployment using local context, only building if necessary (first build or when changes to the local filesystem have been detected)

```
func deploy --build
```
Perform a deployment using local context, forcing a build even if the source is unchanged (useful for example if the user manually removed the resultant container). 

```
func deploy --build=false
```
Perform a deployment using local context, skipping build even if filesystem change detected

```
func deploy --remote
```
Trigger a remote deployment of the local Function by uploading source to be built and deployed remotely.
(Until such time as the local source upload usecase is implemented, the first invocation will error that only --remote with a likewise remote --git-url for source code is supported).

Ideally, a future enhancement would further also only rebuild if the head of the git repository indicated (including branch) has advanced since the last deployment, but that is out of scope for this issue.

```
Func deploy --remote --build
```
Trigger a remote deployment.  Rebuilds the container even if the repository has no new updates.
Presuming the Function has already been configured with an assoicated remote --git-url, this is the way the system currently works (done).

```
func deploy --remote --build=false
```
Trigger a remote deployment.  Reuses the container from the last deployment if it exists.  Essentially equivalent to “redeploy”.

This is also a future enhancement, and could perhaps be implemented without the need of Tekton Pipelines, as it is effectively just re-deploying an existing, built Function.

### The `--git-url` flag

In addition to the –remote and –build dimensions, the –git-url dimension is worth a quick sidebar: we could allow local deploys to pull a git repository in lieu of using the local filesystem, and remote on-cluster deployments could forego using a git repository and instead receive a tarball of local source files.  This dimension is out of scope for this discussion, but it seems a useful thought experiment when deciding to separate the --build flag now rather than later.

Fixes: https://github.com/knative-sandbox/kn-plugin-func/issues/1125

**Comment by zroubalik** [comment]
> ```
> func deploy --remote
> ```
> 
> 
>     
>   
> 
> Trigger a remote deployment.  First invocation requires that --git-url be provided as well.  Will only rebuild if the head of the git repository indicated (including branch) has advanced since the last deployment.
> 

This would be quite tricky to achieve, because we don't have a representation of a function in the cluster (CRD), we would need to do some hacking on previous PipelineRuns? Might be possible though.



> ```
> Func deploy --remote --build
> ```
> 
> 
>     
>   
> 
> Trigger a remote deployment.  Rebuilds the container even if the repository has no new updates.
> 
> ```
> func deploy --remote --build=false
> ```
> 
> 
>     
>   
> 
> Trigger a remote deployment.  Reuses the container from the last deployment if it exists.  Essentially equivalent to “redeploy”.

Again, we would need to probably inspect previous PipelineRuns to achieve this.


> 
> ### The `--git-url` flag
> 
> In addition to the –remote and –build dimensions, the –git-url dimension is worth a quick sidebar: we could allow local deploys to pull a git repository in lieu of using the local filesystem, and remote on-cluster deployments could forego using a git repository and instead receive a tarball of local source files.  This dimension is out of scope for this discussion, but it seems a useful thought experiment when deciding to separate the --build flag now rather than later.

@lkingland I think that it is important to think about `the local source upload` usecase now, when we are designing this big change. What about `--remote=git|upload`?

**Comment by lkingland** [comment]
> > ```
> > func deploy --remote
> > ```
> >   
> > Trigger a remote deployment.  First invocation requires that --git-url be provided as well.  Will only rebuild if the head of the git repository indicated (including branch) has advanced since the last deployment.
> 
> This would be quite tricky to achieve, because we don't have a representation of a function in the cluster (CRD), we would need to do some hacking on previous PipelineRuns? Might be possible though.

Yes I agree.  The second part "only rebuilding if the git repo has changed" is not on the immediate roadmap, but just an example of the logic that would be "complete" if we tried to respect the `--build=auto` for `--remote` builds.

I think the more important thing to notice here is the first part:  triggering a remote deployment without specifying a Git URL as source. Right now by default, this command will generate an error that `--git-url is required".   The first, and much more attainable, next step to realize _this should not be an error case_.   The `func` CLI command operates under the assumption that, by default, every command's context is the source code in the current working directory.  Therefore this command:  `func deploy --remote` should trigger a remote deployment _using the currently active source code_.  That, I believe, is our most urgent (and likely easiest) initial enhancement.

> 
> > ```
> > Func deploy --remote --build
> > ```
> > Trigger a remote deployment.  Rebuilds the container even if the repository has no new updates.
> > ```
> > func deploy --remote --build=false
> > ```
> > Trigger a remote deployment.  Reuses the container from the last deployment if it exists.  Essentially equivalent to “redeploy”.
> 
> Again, we would need to probably inspect previous PipelineRuns to achieve this.

I believe that the first case is how the system currently works, and the second case does not require Tekton.  These are illustrative, and not part of the acceptance criteria for this issue.  I have updated the issue to be more clear on that point.

> @lkingland I think that it is important to think about `the local source upload` usecase now, when we are designing this big change. What about `--remote=git|upload`?

I agree.  The local source upload is the the next most important use case.   As explained above, this case _is actually the default_ `func deploy --remote` = "trigger a remote deployment using this source code".  Right now, since that case it is not implemented, it throws an error.  So, once implemented, the error just goes away and is replaced by the correct behavior:  sending local source to be built and deployed... remotely

**Comment by zroubalik** [comment]
Okay and let say that we have following usecase, how will CLI the experince (commands) look like then?

Let's have a function that uses on cluster build and points to a git repo, so I set the remote type and store the git url in the config. Then I would like to develop the function and would like to quickly iterate and build on the cluster, but I don't want to store these attempts in the git, but I'd like to use direct code upload until I fell comfortable with commiting the change to the git repo. So how do I set temporarily that I would like to use direct upload instead of already configured repo for the on cluster build?

**Comment by lkingland** [comment]
Yes, good question, and I tagged you in a couple of comments to the associated PR.

When I popped the question in chat, people seemed to agree that we should lean towards defaulting to save flags to func.yaml where possible.  So, for example, running `func deploy --git-url ...` would update func.yaml, and subsequent calls to `func deploy` will just use this updated value as the new default.

This of course has positives and negatives, but seems to be more likely to be useful than the opposite of never updating func.yaml and requiring either an explicit call to `func config ...` or manually editing the file.  I am open to further information on this, but so far I agree that persisting flag values by default makes sense.

I say "by deafult", because there of course exists the situation where we want to run a command with specific, _transient_ settings that do not "save" their values into func.yaml.  I don't want to have to manually revert func.yaml.

So if the default is to "save" flag values (where appropriate), then the opposite would of course be `--save=false`.

So, to directly your question, we need to add a new flag which allows the user to explicitly opt out of saving values for flags like `--git-url` for testing such as one-off invocations.

Furthermore, we need to update the concept of "where does the source come from" to be more than just a git url, and also include local paths as well as perhaps future non-git, but network available resources (a tarball somewhere, for example).  For that, we would need to change the `--git-url` flag to be simply `--source`, which would be a git url, file path, or network resource of supported mime type.  It would be nice if we could presume values are a path if they do not contain a scheme, etc.  (I think we should presume file:// because I find it annoying to type.  If I want http, I will always type that!)

```
func deploy --remote --source=. --save=false
```
Build remotely, using the source code from the current local function, but do not save these settings for future invocations)
